mythril_core/src/acpi/hpet.rs

```diff
@@ -0,0 +1,155 @@
+    pub const PAGE_PROTECTION: usize = 19;
+}
+
+/// Page Protection for HPET register access.
+///
+/// See Table 3 in the IA-PC HPET specification.
+#[repr(u8)]
+#[derive(Clone, Copy, Debug, PartialEq, TryFromPrimitive)]
+pub enum PageProtection {
+    /// No page protection
+    NoProtection = 0x00,
+    /// Register access is protected by a 4KB page
+    Protected4KB = 0x01,
+    /// Register access is protected by a 64KB page
+    Protected64KB = 0x02,
+}
+
+/// High Precision Event Timer ACPI entry.
+///
+/// See `IA-PC HPET § 1.0a`.
+pub struct HPET<'a> {
+    /// System Descriptor Table Header for this structure.
+    sdt: &'a SDT<'a>,
+    /// The hardware revision ID.
+    pub hardware_rev_id: u8,
+    /// The number of comparators in the first timer block.
+    pub comparator_count: u8,
+    /// The size cap of the counter. If false, then only 32 bit mode is allowed.
+    /// If true, then both 32 bit and 64 bit modes are supported.
+    pub counter_cap: bool,
+    /// If true, then the HPET is LegacyReplacement IRQ routing capable.
+    pub legacy_replacement: bool,
+    /// The PCI vendor ID of the first timer block.
+    pub pci_vendor_id: u16,
+    /// The address of the HPET register stored as an ACPI GAS.
+    pub address: GenericAddressStructure,
+    /// The HPET sequence number.
+    pub hpet_number: u8,
+    /// The minimum number of ticks that must be used by any counter programmed
+    /// in periodic mode to avoid lost interrupts.
+    pub minimum_tick: u16,
+    /// The type of page protection for HPET register access.
+    pub page_protection: PageProtection,
+}
+
+impl<'a> HPET<'a> {
+    /// Create a new HPET given a SDT.
+    pub fn new(sdt: &'a SDT<'a>) -> Result<HPET<'a>> {
+        let event_timer_block_id =
+            NativeEndian::read_u32(&sdt.table[offsets::EVENT_TIMER_BLOCK_ID]);
+        let address =
+            GenericAddressStructure::new(&sdt.table[offsets::BASE_ADDRESS])?;
+        let hpet_number = sdt.table[offsets::HPET_NUMBER];
+        let minimum_tick =
+            NativeEndian::read_u16(&sdt.table[offsets::MIN_CLOCK_TICK]);
+
+        let page_protection = match PageProtection::try_from(
+            sdt.table[offsets::PAGE_PROTECTION] & 0xF,
+        ) {
+            Some(page_protection) => page_protection,
+            None => {
+                return Err(Error::InvalidValue(format!(
+                    "Invalid HPET Page Protection type: {}",
+                    sdt.table[offsets::PAGE_PROTECTION] & 0xF
+                )));
+            }
+        };
+
+        let hardware_rev_id = (event_timer_block_id & 0xFF) as u8;
+        let comparator_count = ((event_timer_block_id & 0x1F00) >> 8) as u8;
+        let counter_cap = ((event_timer_block_id & 0x2000) >> 13) as u8;
+        let legacy_replacement = ((event_timer_block_id & 0x8000) >> 15) as u8;
+        let pci_vendor_id = ((event_timer_block_id & 0xFFFF0000) >> 16) as u16;
+
+        let counter_cap = match counter_cap {
```

Comment:
I'd rather write this as `let counter_cap = counter_cap != 0;`. Or possible just do this during the initial assignment to `counter_cap`. Same for `legacy_replacement` below.